### PR TITLE
AP-9818 # Added calculationService

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **Medium Risk**
> Payment amount resolution now depends on expression evaluation and new amount-type configuration; incorrect expressions or coercion could affect charged amounts, though behavior is covered by new tests.
> 
> **Overview**
> Adds **`calculationService.evaluateExpression()`** to evaluate OneBlink calculation strings against submission data via the new **`morph-expressions`** runtime dependency. It resolves `{ELEMENT:...}` references (including nested/repeatable paths), registers **`ROUND`**, **`ROUND_DOWN`**, **`ROUND_UP`**, and **`ISNULL`**, and uses caller-injected **`parseDayOnlyDate`** for `YYYY-MM-DD` values (shared **`DATE_ONLY_PATTERN`** moved to conditional logic types).
> 
> **`paymentService.checkForPaymentEvent()`** no longer assumes a single form element for the amount: **`resolvePaymentEventAmount()`** branches on **`amountType`** — fixed **`paymentAmount`**, **`paymentCalculation`** (expression), or **`elementId`** (form element) — with existing validation for zero/non-numeric amounts.
> 
> The package exports **`calculationService`**, adds TypeScript declarations for **`morph-expressions`**, and extends unit tests for calculations and payment event amount resolution.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c1451f9c32467ef7112c7bee99167d27ad17b5ad. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/calculationService.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    if (!Number.isNaN(value) && Number.isFinite(value)) {
+      return Math.floor(value)
+    }
+    return null
+  })
+  exprParser.registerFunction('ROUND_UP', (value: number) => {
+    if (!Number.isNaN(value) && Number.isFinite(value)) {
+      return Math.ceil(value)
+    }
+    return null
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/calculationService.ts:341]: This is current implementation copied from apps-react. I would prefer not to change it.

File: src/calculationService.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+      if (typeof elementValue === 'object') {
+        const formFormElement = findFormElement(
+          formElements,
+          (e) => e.type === 'form' && e.name === nestedElementNames[index],
+        )
+        const nextElementName = nestedElementNames[index + 1]
+        if (formFormElement && nextElementName) {
+          return (elementValue as Record<string, unknown>)[nextElementName]
+        }
+      }
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/calculationService.ts:154]: This is current implementation copied from apps-react. I would prefer not to change it.

File: src/paymentService.ts [hunk 3]
```diff
@@ -150,12 +140,79 @@ export function checkForPaymentEvent({
...
+          'We could not find the configuration required to make a payment. Please contact your administrator to ensure your application configuration has been completed successfully.',
+        )
+      }
 
-  return result
+      const amountElement = formElementsService.findFormElement(
+        definition.elements,
+        (element) => element.id === configuration.elementId,
+      )
+      if (!amountElement || amountElement.type === 'page') {
```
Comment (divporter) [inline src/paymentService.ts]: I know it's from apps-react and I probably have the same line in my PR but there are more invalid types that page. Do you think we should tighten this up?

**Review by divporter — APPROVED** [review]

File: src/paymentService.ts (hunk 3 above)
Comment (mymattcarroll) [reply src/paymentService.ts]: I saw this as well but figured changing it is not worth it. I am board with the removing the check for `page` element types as you say there are other invalid types. I believe this was here were we requiring an element with a `name` property in the next line(s) but is no longer the case.

```suggestion
      if (!amountElement) {
```
